trecs/models/recommender.py

```diff
@@ -337,22 +348,69 @@ def generate_recommendations(self, k=1, item_indices=None):
             negated_scores = -1 * s_filtered  # negate scores so indices go from highest to lowest
-            top_k = negated_scores.argpartition(k - 1)[:, :k]
+            # break ties using a random score component
+            scores_tiebreak = np.zeros(
+                negated_scores.shape, dtype=[("score", "f8"), ("random", "f8")]
+            )
+            scores_tiebreak["score"] = negated_scores
+            scores_tiebreak["random"] = self.random_state.random(negated_scores.shape)
+            top_k = scores_tiebreak.argpartition(k - 1, order=["score", "random"])[:, :k]
             # now we sort within the top k
             row = np.repeat(self.users.user_vector, k).reshape((self.num_users, -1))
             # again, indices should go from highest to lowest
-            sort_top_k = negated_scores[row, top_k].argsort()
+            sort_top_k = scores_tiebreak[row, top_k].argsort(order=["score", "random"])
             rec = item_indices[
                 row, top_k[row, sort_top_k]
             ]  # extract items such that rows go from highest scored to lowest-scored of top-k
             if self.is_verbose():
-                self.log(f"Row:\n{str(row)}")
                 self.log(f"Item indices:\n{str(item_indices)}")
                 self.log(
-                    f"Top-k items ordered by preference (low to high) for each user:\n{str(rec)}"
+                    f"Top-k items ordered by preference (high to low) for each user:\n{str(rec)}"
                 )
             return rec
 
+    def choose_interleaved_items(self, k, item_indices):
+        """
+        Chooses k items out of the item set to "interleave" into
+        the system's recommendations. **NOTE**: Currently, there
+        is no guarantee that items that are interleaved are distinct
+        from the recommended items. We do guarantee that within the
+        set of items interleaved for a particular user, there are no
+        repeats.
+
+        Parameters
+        -----------
+
+            k : int
+                Number of items that should be interleaved in the
+                recommendation set for each user.
+
+            item_indices : :obj:`numpy.ndarray`
+                Array that contains the valid item indices for each user;
+                that is, the indices of items that they have not yet
+                interacted with.
+
+        Returns
+        ---------
+            interleaved_items: :obj:`numpy.ndarray`
+        """
+        if self.interleaving_fn:
+            return self.interleaving_fn(k, item_indices)
+
+        if k == 0:
+            return np.array([]).reshape((self.num_users, 0)).astype(int)
+
+        # NOTE: there is currently no guarantee that randomly interleaved items do
+        # not overlap with recommended items, since we do not have visibility
+        # into the recommended set of items. we do guarantee that for every user,
+        # items will not be repeated within the set of interleaved items.
```

Comment:
This seems non-optimal. Why aren't we just running like a `set` on the recommendations at each timestep to avoid this?

Reply:
Because if you want to show a constant number of items to each user, then you'd have to "fill" the remaining recommendations with new items that are sampled...somehow (potentially with a slow loop). It also doesn't really matter *too too* much unless there are item ordering effects.

In the future, it'd be very possible to stop the repeat phenomenon by just setting the score sort value to negative infinity for items that are going to be recommended...I'll open an issue to resolve that!